### PR TITLE
Modernize ironbank Dockerfile

### PR DESCRIPTION
**Pull request by kaisecheng**
- remove golang assets (go1.17.8.linux-amd64.tar.gz)
- remove yaml lib assets (v2.3.0.tar.gz)
- use go container to build env2yaml
- remove unnecessary layers
- remove HEALTHCHECK
- switch yum to dnf

I have made some changes from the suggested Dockerfile
- remove `COPY v2.3.0.tar.gz /tmp/env2yaml.tar.gz`
- remove `mkdir -p /usr/local/src/vendor/gopkg.in && \
    tar -zxf /tmp/env2yaml.tar.gz -C /go && \`
- add `mv config/* /usr/share/logstash/config && \
  mv pipeline /usr/share/logstash/pipeline && \`
- add `rmdir config && \`

Fixes: https://github.com/elastic/ingest-dev/issues/3008

## Author's Checklist
- [ ] checked files and folders have the same permission as the old container, `/usr/share/logstash`, `/usr/share/logstash/config`, `/usr/share/logstash/pipeline`, `/usr/local/bin`
- [ ] checked the image can [run](https://repo1.dso.mil/dsop/elastic/logstash/logstash/-/jobs/32812669) gracefully by trying`RUN "/usr/local/bin/docker-entrypoint" -e "input{ generator{count => 1} } output{ stdout{}}"`

## How to test this PR locally

- run `rake artifact:dockerfile_ironbank`
- take `Dockerfile` and `hardening_manifest.yaml` from logstash/build/logstash-ironbank-8.14.0-SNAPSHOT-docker-build-context.tar.gz
- create a test PR in GitLab with those two files. CI should pass. You may need to replace `8.14.0-SNAPSHOT` with `8.12.2` and corresponding checksum in `hardening_manifest.yaml` to test

### CODE REVIEW

File: docker/templates/hardening_manifest.yaml.erb [hunk 1]
```diff
@@ -44,16 +44,6 @@ resources:
   validation:
     type: sha512
     value: <INSERT SHA512 VALUE FROM https://artifacts.elastic.co/downloads/logstash/logstash-<%= elastic_version %>-linux-x86_64.tar.gz.sha512>
-- filename: go1.17.8.linux-amd64.tar.gz
-  url: https://dl.google.com/go/go1.17.8.linux-amd64.tar.gz
-  validation:
-    type: sha256
-    value: 980e65a863377e69fd9b67df9d8395fd8e93858e7a24c9f55803421e453f4f99
-- filename: v2.3.0.tar.gz
```
Comment (kaisecheng) [inline docker/templates/hardening_manifest.yaml.erb:52]: We build env2yaml with `go build` now, while the old way is `go build -mod vendor` taking the source from vendor.
v2.3.0.tar.gz is no longer need

**Comment by robbavey** [comment]
@kaisecheng I wonder if at this point, it makes sense to split this up into two separate Dockerfile templates, ironbank and elastic? The Ironbank Dockerfile is almost completely different to the elastic one at this point, and while we might converge again, it's quite difficult to tell at a glance which distribution a line of code relates to.

WDYT?

**Comment by kaisecheng** [comment]
I think the new commit is easier to follow comparing to the last one, which messed up with all image flavours without a clear border.
Now ironbank and elastic share the last four layers, and I have a feeling that all layers will be merged again soon because all the modernization happens here can apply to elastic and the Go related stuffs will be removed in other task (#15766). So, keeping them in a single file to me is more intuitive.

File: docker/templates/Dockerfile.erb [hunk 2]
```diff
@@ -1,148 +1,147 @@
...
-<% elsif image_flavor == 'ironbank' -%>
-<%   package_manager = 'yum' -%>
-<% else -%>
-<%   base_image = 'ubuntu:20.04' -%>
-<%   package_manager = 'apt-get' -%>
-<%   locale = 'en_US.UTF-8' -%>
-<% end -%>
 <% if image_flavor == 'ironbank' -%>
-ARG BASE_REGISTRY=registry1.dsop.io
+ARG BASE_REGISTRY=registry1.dso.mil
```
Comment (robbavey) [inline docker/templates/Dockerfile.erb:3]: Can we indent the `ironbank` image definition?

File: docker/templates/Dockerfile.erb [hunk 3]
```diff
@@ -1,148 +1,147 @@
...
+  chmod -R g=u /usr/share/logstash && \
+  mv config/* /usr/share/logstash/config && \
+  mv pipeline /usr/share/logstash/pipeline && \
+  mkdir /licenses/ && \
+  mv /usr/share/logstash/NOTICE.TXT /licenses/NOTICE.TXT && \
+  mv /usr/share/logstash/LICENSE.txt /licenses/LICENSE.txt && \
+  ln -s /usr/share/logstash /opt/logstash && \
+  chmod 0755 /usr/local/bin/docker-entrypoint && \
+  rmdir config && \
+  rm /tmp/logstash.tar.gz
```
Comment (robbavey) [inline docker/templates/Dockerfile.erb:57]: Can we add a comment to state that this closes the `<% if image_flavor == 'ironbank' -%>`

File: docker/templates/Dockerfile.erb [hunk 4]
```diff
@@ -1,148 +1,147 @@
...
+    <%   base_image = 'docker.elastic.co/ubi8/ubi-minimal' -%>
+    <%   package_manager = 'microdnf' -%>
+    # Minimal distributions do not ship with en language packs.
+    <%   locale = 'C.UTF-8' -%>
+  <% else -%>
+    <%   base_image = 'ubuntu:20.04' -%>
+    <%   package_manager = 'apt-get' -%>
+    <%   locale = 'en_US.UTF-8' -%>
+  <% end -%>
+
```
Comment (robbavey) [inline docker/templates/Dockerfile.erb:82]: Can we indent the code under here to indicate that it is part of the `else <% if image_flavor == 'ironbank' -%>` block?

File: docker/templates/Dockerfile.erb [hunk 5]
```diff
@@ -155,18 +154,9 @@ ARG TARGETARCH
...
-COPY scripts/config/log4j2.properties config/
-COPY scripts/config/log4j2.file.properties config/
-COPY scripts/pipeline/default.conf pipeline/logstash.conf
-RUN chown --recursive logstash:root config/ pipeline/
-# Place the startup wrapper script.
-COPY scripts/bin/docker-entrypoint /usr/local/bin/
-<% end -%>
 
 RUN chmod 0755 /usr/local/bin/docker-entrypoint
+<% end -%>
```
Comment (robbavey) [inline docker/templates/Dockerfile.erb:159]: Can we comment that this marks the end of the `else <% if image_flavor == 'ironbank' -%>` block?

**Review by robbavey — CHANGES_REQUESTED** [review]
Request for some changes to help clarify where ironbank and non-ironbank Dockerfile code resides

File: docker/templates/Dockerfile.erb (hunk 2 above)
Comment (kaisecheng) [reply docker/templates/Dockerfile.erb:3]: Do you mean add \t to all lines inside `if image_flavor ==  'ironbank'` block?
The output Dockerfile will result in indentation as well. I think we don't want it, isn't it?

**Comment by kaisecheng** [comment]
@robbavey I added comments for start and end of block. Keep in mind that the comments will show in the output Dockerfile

**Comment by robbavey** [comment]
@kaisecheng - you can add comments in .erb syntax that don't show up in the Dockerfile

```
<%# This is an ERB comment and won't appear in the output %>
```

**Review by robbavey — APPROVED** [review]
I've kicked off exhaustive tests, which include the docker build.

https://buildkite.com/elastic/logstash-exhaustive-tests-pipeline/builds/339
https://buildkite.com/elastic/logstash-aarch64-pipeline/builds/78

Assuming these are green, LGTM

**Comment by kaisecheng** [comment]
@logstashmachine backport 8.13

**Comment by kaisecheng** [comment]
@logstashmachine backport 8.13

**Pull request merged**